tango/local_workspace.py

```diff
@@ -275,9 +275,38 @@ def work_dir(self, step: Step) -> Path:
         result.mkdir(parents=True, exist_ok=True)
         return result
 
-    @staticmethod
-    def dir_is_empty(dir: Path):
-        return not any(True for _ in dir.iterdir())
+    @classmethod
+    def guess_step_dir_state(cls, dir: Path) -> Optional[StepState]:
+        """
+        Returns the status of a given step dir, to the best of our knowledge.
+
+        :param dir: the step dir to example
+        :return: our best guess at the state of the step, or ``None`` if we don't know
+        """
+
+        # If the directory doesn't exist, the step is incomplete.
+        if not dir.exists():
+            return StepState.INCOMPLETE
+
+        # If the lock file exists and is locked, the step is running.
+        lock_file = dir / "lock"
+        if lock_file.exists():
+            lock = FileLock(lock_file)
+            try:
+                lock.acquire(0)
```

Comment:
I think it would be a little safer to acquire with a small timeout, like 0.2 for example. That would help avoid the situation where two (or more) separate Python processes reach this line of code at the same time, both trying to acquire the lock immediately. Because only one will succeed, and the other(s) will erroneously think the step is running.

Reply:
8d8785f